### DISCUSSION TITLE
How to use only security keys as 2fa?

### DISCUSSION THREAD

**Discussion by alex-meunier** [community/community#22500] [New to GitHub]
<p>Hi,</p>
<p>I have two security keys and I would want it to be the only source of 2fa for my account but Git Hub still allows me to authenticate using text messages or recovery codes. It seems I can’t manage these options as I would like to. Moreover this documentation <a href="https://docs.github.com/en/authentication/securing-your-account-with-two-factor-authentication-2fa/configuring-two-factor-authentication#configuring-two-factor-authentication-using-fido-u2f" class="inline-onebox" rel="noopener nofollow ugc">Configuring two-factor authentication - GitHub Docs</a> says</p>
<blockquote>
<p>Authentication with a security key is <em>secondary</em> to authentication with a TOTP application or a text message. If you lose your security key, you’ll still be able to use your phone’s code to sign in.</p>
</blockquote>
<p>I could accept to keep the recovery codes option, but I don’t want text messages or TOTP application. Can I configure 2fa like this ?</p>
<p>Thanks for reading me :blush:</p>

**Comment by joshuawalker** [comment, marked as answer]
<p>Hey <a class="mention" href="/alex-meunier">@alex-meunier</a>, we intend to support using a security key as the initial two-factor authentication method without the need to first setup an authenticator app or SMS number, but we do not have a timeframe for when this will be implemented. In the meantime, you’ll have to configure a TOTP app or SMS number before you can use a security key.</p>

**Reply by frakier** [reply]
so unless you have a cell phone permanently attached to your rear you are out of luck.

**Reply by candymappr** [reply]
Pretty much

**Reply by frakier** [reply]
Was on a leash for work for a long time, got rid of that thing. I'm reminded as much anytime a bunch of friends/my brother are standing around a campfire, its all peaceful, middle of nowhere and all their phones start ringing as their wives/husband, girlfriends/boyfriends, kids, bosses, etc. start yanking that chain... and then they all answer knowing it is going to completely ruin the rest of their weekend and whatever wilderness class we have scheduled becomes work instead of fun. Github should have made something, even just signed up with something like privacy pass instead of a half way attempt at implementation... just deleted my repositories before i get locked out.

**Reply by Ayzzzzzzzzzzz** [reply]
yes

**Reply by yaxxie-piramidal** [reply]
> but we do not have a timeframe for when this will be implemented.

its been three years since this was stated and its becoming an unacceptable timeframe with still no timeframe in sight. As others have stated, I want to lose my account if i lose my security keys. Even worse, you're forcing me to periodically use the TOTP even when my security key is present! This is madness!

**Comment by mars-risc0** [comment]
Any progress on this? I definitely do not ever want to use my phone for authentication, only the security key.

**Reply by haifeng-jin** [reply]
I am also looking forward to this feature.

**Reply by sherppard** [reply]
same as me, we need this option.

**Reply by sherppard** [reply]
@joshuawalker

**Reply by AndrewMohawk** [reply]
+1 From me

**Reply by clwgh** [reply]
+1 from me -- I don't want a fallback of TOTP or SMS for a hardware key because they short-circuit the whole point of using a hardware key.

**Reply by philphilphil** [reply]
+1

**Reply by liamriley1** [reply]
Please fix this - forcing me to add another, worse option makes the security key less secure and is just a stupid idea in general

**Reply by ChrisHills463** [reply]
Please add me to the list of users who wants to only permit the use of security keys to access my account (other than the recovery codes).

**Reply by danpaldev** [reply]
Having a key is useless with this stupid limitation of yours. If I'm still depending on the worthless TOTP what's the point of having a key? We're not 5-years old who need to be kept away from hurting themselves.

**Reply by tmanley1985** [reply]
I too only want to use yubikeys for 2fa.

**Reply by kevinforrestconnors** [reply]
Me too

**Comment by DNin01** [comment]
GitHub is working on making it possible to use security keys as the primary second factor, based on what they said in [a blog post](https://github.blog/changelog/2022-11-21-updates-to-the-two-factor-authentication-setup-flow/).
> In the future, you'll also find security keys there as an option for initial setup on supported devices and browsers.

**Comment by emlun** [comment]
It's great to see GitHub rolling out the 2FA requirement for active contributors!

When will we be able to choose _only_ security keys and passkeys for sign-in to GitHub? In other words:

- I want to ***not allow SMS*** for 2FA to my account,
- I want to ***not allow TOTP*** for 2FA to my account,
- I want to ***not allow GitHub Mobile*** for 2FA to my account,
  - (but I do want to be able to use the mobile app, just not allow it as a 2FA credential for sign-in on other devices)
- I want to ***not allow recovery codes*** for 2FA to my account, and
- I ***want to be locked out*** if I lose all ≥2 of my security keys.

It is already the case in practice that my security keys are the only way I can authenticate to GitHub (except for GitHub Mobile). My TOTP credential is stored on the same YubiKeys as I use exclusively as security keys. My recovery codes are stored in my password vault, which is locked with the same YubiKeys as I use exclusively as security keys. Ergo, forcing me to set up TOTP and recovery codes does not make me less likely to lose access to my account. All it does is increase my attack surface. I'm certain this is true of many other users as well.

I'm not asking GitHub to drop support for these authentication methods. But please allow me to delete them from my account if I want to.

Though I guess as a workaround, I could set up a new TOTP key and generate new recovery codes, and immediately destroy both instead of saving them. However that still doesn't eliminate GitHub Mobile as a downgrade attack vector.

**Reply by danpaldev** [reply]
Gitea already allows this, so I moved there. I won't be pushing anything to Github until they fix this.

**Reply by joshuawalker** [reply]
Thanks for your feedback! We don't have a date to share for when security keys will be able to be used as an initial 2FA method. We'll share news on this on the [GitHub Changelog](https://github.blog/changelog/).

**Reply by clwgh** [reply]
Thanks Joshua, please put focus on this in order to fix the security vulnerabilities with your current implementation as quickly as possible. The original post is now 13 months old with no fix or even an ETA available.

**Reply by MightyBOBcnc** [reply]
> Thanks for your feedback! We don't have a date to share for when security keys will be able to be used as an initial 2FA method. We'll share news on this on the [GitHub Changelog](https://github.blog/changelog/).

It's ridiculous that *GitHub*, of all websites, doesn't allow hardware keys as a primary/sole 2FA method, especially when it's pushing mandatory 2FA enrollment.

**Reply by candymappr** [reply]
I won't be able to log in because I don't have a phone.

**Reply by DNin01** [reply]
> I won't be able to log in because I don't have a phone.

Maybe there's an authenticator app that would work on your computer? You could even use a password manager with a TOTP feature.

**Reply by candymappr** [reply]
Thanks so much for the idea!
I found winauth and i can now use github.
        Thank you

**Reply by mrniko** [reply]
As option you can use Yubico Authenticator. It doesn't require a phone number or internet access. Secret key will be stored in your hardware YubiKey and not locally. Don't forget that the local time on your device should be synced with global time.

**Reply by candymappr** [reply]
thanks @mrniko I now have a system i like

**Comment by Perkolator** [comment]
I just got the "Your GitHub account will soon require 2FA" email and I don't care about 2/MFA. I have very secure passwords and have them secured very well.

The email mentioned this:

>  We want you to have the most seamless experience with 2FA possible, so you can choose one or more of the following options:
>
>   - Security key
>   - GitHub Mobile
>   - Authenticator application (TOTP)
>   - Text messages (SMS)

But I can see that only TOTP and SMS is offered when enabling 2FA. I could use security keys (would go to same password vault as the actual password for the account.. doesn't make any sense, but if that gets around this github 2FA stuff, great). Otherwise have to think about leaving for another solution.

**Reply by alex-meunier** [reply]
Very strong password are not sufficient anymore for great account security. Today, too many accounts are being hacked on all websites, even if they have strong passwords because there are much more powerful hacking techniques than just brut forcing their passwords. Even strong password + 2fa with SMS isn't enough either, as huge security issues have been discovered on this type of 2fa. Currently, security keys are the safest. That's why everyone on this thread wants to use only security keys and not SMS. I guess GitHub won't let you the choice to log in with only a password because all the websites are currently enabling 2fa for all accounts and it will become more and more common in the future.

**Reply by Perkolator** [reply]
> all the websites are currently enabling 2fa for all accounts

Enabling Vs. forcing. I have over 100 accounts in my password vault and github is the first that is forcing me to use 2/MFA (except my bank, which is welcomed and somewhat easy with the bank app or code list, and I really don't have to worry about locked out of my bank account, unlike with 2/MFA with other sites.. I don't like the pitfalls of 2/MFA).

> and it will become more and more common in the future.

Will see.. at least when current encryption stuff is broken with quantum computing.

> too many accounts are being hacked on all websites, even if they have strong passwords because there are much more powerful hacking techniques than just brut forcing their passwords

Like what? Stolen DB and rainbow tables?

> Currently, security keys are the safest.

After reading the github email, that's what I was trying to use, but it's not available. They wanted me to have a "most seamless exp" with choosing "one or more of the following options" listed in the email... talk about false advertising.

**Reply by liamriley1** [reply]
I don’t think it’s a “much more powerful” hacking technique, but I would say that one of the key advantages of security keys over very long passwords is the lack of opportunity for phishing attacks. With FIDO, there really is only your security key that can unlock your accounts, and the credentials cannot be captured by bad actors. The numbers speak for themselves when you look at a company like Google, who switched their employees to security keys in 2017 and has reported 0 phishing incidents since then. I agree that it seems extremely counterintuitive to GitHub to start this huge drive for users to configure 2FA, but then force them to use slower and less secure methods

**Reply by DNin01** [reply]
The closest you could probably get is either using Yubico Authenticator (which unlocks with a security key) or even deleting the code from authenticator app if you wanted to do that, then also adding security keys.

**Reply by DNin01** [reply]
Some password managers also allow saving TOTP codes, so if you use one, as long as the vault also uses 2FA, that would be a good workaround.

**Reply by liamriley1** [reply]
> The closest you could probably get is either using Yubico Authenticator (which unlocks with a security key) or even deleting the code from authenticator app if you wanted to do that, then also adding security keys.

You'd think that you could just delete the TOTP code from your authenticator app right? But NO, a few weeks later GitHub will make you authenticate via the TOTP code and won't let you login to GitHub (even with your perfectly valid security key) until you reconfigure TOTP to work again and give it a valid code. It seems GitHub has chosen the approach of babying their users and assuming that they have set their authentication up wrong. I just want be locked out if my security keys don't work, and I don't have one of my backup codes. Should be a very simple idea but all of these stipulations that GitHub has come up with make it so unnecessarily complicated!

**Reply by ghost** [reply]
I agree that we should be able to deactivate TOTP if we have set up **more than one** security key.

**Reply by MasterInQuestion** [reply]
“Very strong password are not sufficient anymore for great account security.
    Today, too many accounts are being hacked on all websites, even if they have strong passwords because there are much more powerful hacking techniques than just brut forcing their passwords.”
    Password itself may function as if security key:
    "/dev/random" derived, unique for each site.
    When properly managed: much identical. (and likely more reliable)

    See also: https://github.com/orgs/community/discussions/63813#discussioncomment-10626201
\
\
    It appears many people have fundamental flaws on the understanding of security:
    Thoughtlessly phasing things out in vain hope, for feigned sense of "security".
    But failed to realize:
    “Security doesn't itself come by following random security advices or blindly deploying "security-enhanced" setup:
    Well realizing the underlying implementation is the essence to guarantee effectiveness and security.”
    .
    https://github.com/orgs/community/discussions/22814#discussioncomment-7571079

**Reply by Explorer09** [reply]
@MasterInQuestion Why are we believing in a bulls**t that "Password itself may function as if security key"?

Passwords, no matter how random or secure you generated them, are the same tokens every time you log in. Passwords are vulnerable to [replay attacks](https://en.wikipedia.org/wiki/Replay_attack). That was the reason for 2FA in the first place. Applications that have to stick with password authentication are shifted to use [personal access token](https://docs.github.com/en/authentication/keeping-your-account-and-data-secure/managing-your-personal-access-tokens) as a password alternative. Personal access tokens can expire after a set days, and have limited privileges per token, and thus reduce the risks of what passwords can do.

**Reply by MasterInQuestion** [reply]
Before bold assertion: do proper research.
    Why would you fallaciously believe that random keys each time would miraculously increase security?
    For some inconsiderate genius coined that the ideal encryption model would be alike, given the "absolute reliable" key-exchange channel..?

    For really security-critical applications: 1 compromise already means done.
    Reusable or not is already passé.

    Your thoughts are exactly like those who coined security models that don't work at all:
    Secure or not: much in question.
    Functionality-wise: everything broken.

    Your quoth of Replay Attack is also not much appropriate:
    But whatsoever doesn't matter much.

**Reply by Explorer09** [reply]
@MasterInQuestion 
> Why would you fallaciously believe that random keys each time would miraculously increase security?

I don't. I am just refuting your claim about passwords being secure enough.
If what you claim is true, why the heck do websites force 2FA on us?

**Reply by MasterInQuestion** [reply]
Thoughtless assertion that users are all thoughtlessly using literally "password" as password.
    Some did alike, indeed.
    But such apparently don't care about security.
    Mostly nor capable of maintaining security.
    Critical business demands operator competence: an unavoidable fact.

    Good faith assumed, however:
    Some just excuse this for data collection.

**Reply by Explorer09** [reply]
@MasterInQuestion

There are a lot of password strength checkers you can find on the web. Please don't assume we know nothing about password strength.
* https://www.security.org/how-secure-is-my-password/
* https://bitwarden.com/password-strength/
* https://www.passwordmonster.com

For your argument about data collection, yes, this is my own reason **not to use 2FA approach other than the security key**. SMS based 2FA leaks your phone number. App based 2FA (including TOTP) requires you to trust the app -- something can be programmed to spy on you. Hardware security keys stay offline until you need them for authentication, so it is the best 2FA approach for people who have privacy concerns.

**Reply by MasterInQuestion** [reply]
I don't. Find who force 2FA to blame.
    Hardware-based are not necessarily safer than software.
    And actually hardware and software cannot really separate...

    The key benefit of security keys alike: is the key generation core being not easily copiable from outside.
    In some cases this may of help.
    But do note: operating security keys in untrusted environment still presents the risk of session hijacking.

**Reply by liamriley1** [reply]
Just to add to the discussion about SMS 2FA: not only is it collecting even more user data but it is shockingly easy to hack an account with SMS 2FA. A simple social engineering attack (for example calling the user's phone company and asking for a new SIM card for that phone number) will easily bypass the security and there are many documented examples of that specific "SIM swap" attack happening in real life. 

It's 2024 and this is an embarrassment. Nearly every other big name in the software versioning and collaborating area respects the users decision of 2FA method and doesn't force vastly less secure methods. All kinds of mission-critical applications and products could be affected by a compromised Github account, and this has the potential to cause large disruption to the general public who rely on these products. Continuing this security stance is dangerous and irresponsible. Please fix this asap.

**Reply by Explorer09** [reply]
@MasterInQuestion 

> Find who force 2FA to blame.

GitHub.

> is the key generation core being not easily copiable from outside.

Hardware security keys rely on physical security. Yes, the key themselves can be cloned if they are physically stolen. But remote stealing of security key data is very unlikely (because NFC requires a very close range (a few centimeters) to communicate). As long as I physically hold my keys (thus having control on the physical security), there is no security weakness with the keys.

**Reply by MasterInQuestion** [reply]
Depends how you use it.
    Eventually fallback to the key problem: key management.

    Note what adversary targets is your access: not necessarily you keys.

**Reply by Explorer09** [reply]
@MasterInQuestion Your argument was about passwords being as strong as the security key. While we refute your arguments, you are now dodging questions.

Don't propose any substitute for security keys while you know little about security keys. This whole thread is about "we want security keys as an exclusive second factor; no TOTP", and discussions about password strengths are already off-topic. Please don't waste our time.

**Reply by MasterInQuestion** [reply]
I don't use it doesn't mean I don't understand the technicality behind.
    The point was not on keys themselves, but security.

    Note my original "much identical", not "identical".
    I've explained the primary differences.

    On-topic with the adjacent posts.
    Per main topic OT in deed.
    Exported: [ https://github.com/MasterInQuestion/talk/discussions/38](https://github.com/MasterInQuestion/talk/discussions/38)

**Reply by Explorer09** [reply]
@MasterInQuestion You are assuming people's accounts get hacked only because of weak passwords. They aren't. The servers' database can get compromised, leaking all of the password hashes of their users. Long and random passwords can still get defeated by a keylogger if users log in in insecure machines. There are never chances that passwords would be "secure enough". That's what 2FAs are for, d***it.

**Reply by MasterInQuestion** [reply]
I'd suggest reading previous contexts carefully.

**Comment by John-Nagle** [comment]
Has Github dropped support for Yubikeys? The link on how to do this is now a dead link:

_[Here is](https://resources.github.com/whitepapers/GitHub-guide-to-commit-signing/) a detailed configuration guide for setting up your YubiKey with GitHub for commit verification and for SSH-based authentication._

**Reply by diegoescala** [reply]
Looks like they require you to use an authenticator app first. So you'll need your phone no matter what.

**Reply by DNin01** [reply]
An authenticator app or phone number is required during the initial 2FA setup process and must be available as an option if you want to use 2FA. One option that pairs with a security key is [Yubico Authenticator](https://www.yubico.com/products/yubico-authenticator/), which supports both mobile and desktop platforms.

Once you have set that up, you can use your security key for 2FA as well.

**Reply by clwgh** [reply]
"Once you have set that up, you can use your security key for 2FA as well."

As well, but not exclusively, which means there is always a lower-security bypass for the high-security security key, which is the complaint being made, and ignored by GitHub for 2 years now, in this thread.

**Reply by joshuawalker** [reply]
@clwgh We're not ignoring the feedback here. In fact, since my last comment in this Discussion, we announced that [we will support security keys](https://docs.github.com/en/authentication/securing-your-account-with-two-factor-authentication-2fa/about-mandatory-two-factor-authentication#about-required-2fa-methods) as a primary 2FA method once we're comfortable with the level of sync support and usability generally. Additionally, [we began supporting passkeys](https://github.blog/2023-07-12-introducing-passwordless-authentication-on-github-com/) which allows you to upgrade a security key to a passkey and use it for both password and two-factor authentication simultaneously.

**Reply by Explorer09** [reply]
@joshuawalker Your comment is annoying. Since you guys are going to force 2FA to many users (https://www.bleepingcomputer.com/news/software/github-warns-users-to-enable-2fa-before-upcoming-deadline/), why the heck can't you even get the sense of security right?

People who care about security and privacy don't want TOTP as a fallback. Those who own security keys would have two keys or more to register at the same time, so losing one key isn't a concern.

And we don't want passkeys either. Passkeys are a hell of inconvenience for us who frequently log in without a phone or any device syncing. Just give as the password + U2F security key as the option. No excuses, please.

**Reply by clwgh** [reply]
@joshuawalker The article you linked to does not "announce that you will support security keys as primary 2FA" at all. Instead it refers to eventually doing this for "passkeys", which is a broad term that equally covers mobile phone authentication with fingerprints or FaceID. This makes me think that you will use those methods instead.

Will you please allow people to use their YubiKeys, and similar hardware security keys, as primary 2FA? You could require that 2 or more keys be used. During setup you could offer an optional one-time set of 10 emergency access codes. Let the users, who are savvy enough to use and understand hardware keys, make the decisions about the security of their accounts.

It's been 2+ years in the asking, so you can forgive the perception that GitHub absolutely is ignoring feedback. Your current stance is akin to offering a high security front-door lock as long as a spare key is hidden in a plant pot on the front step.

**Reply by Explorer09** [reply]
@clwgh @joshuawalker Just to give a note that "passkeys" and "security keys as 2FA" are two different things. While the industry is slowly accepting passkeys as an alternative to password login. Passkeys are not a good choice for everyone.

And be careful that passkeys ≠ 2FA. Passkey is only a single factor authentication while the passkey-supported device (your smartphone or laptop) may provide an optional second factor (PIN, face recognition, fingerprint sensor, etc.)
The keyword is _optional_, so passkey is **not** a 2FA by itself.

The main drawback of using passkeys is that you have to sync the secrets across devices. That reason alone makes it not a good replacement for passwords, for certain people.

For me personally, I would just rely on a strong password, and YubiKeys for the occasional second factor auth. I don't trust other methods.

**Reply by clwgh** [reply]
I get the impression that GitHub wants to end up supporting some variant of passkeys as a primary 2FA approach, and this will allow GitHub to market the new account feature, while avoiding having to manage the process for loss of 2FA, instead handing that work over to the passkeys provider.

Watch as lots of providers take this third-party hands-off approach to account security over the next couple of years.

**Reply by Explorer09** [reply]
@clwgh Trusting passkeys as a "primary 2FA" is a terrible idea. Passkeys means that you have to always bring the device (phone) with you whenever you log in. I know YubiKeys can be configured as passkeys, but I won't do that for the very reason that I don't want to bring up my key every time I log in (but only doing privileged tasks would require the 2FA).
The convenience factor of passwords is irreplaceable.

**Reply by joshuawalker** [reply]
@clwgh @Explorer09 GitHub will support using a security key as the initial two-factor authentication method without the need to setup a TOTP app, SMS number, or passkey. We do not have a timeframe for when this will be supported. We will share changes at the [GitHub Changelog](https://github.blog/changelog/).

**Reply by clwgh** [reply]
Thanks for the comment @joshuawalker, it is reassuring that proper keys will eventually be supported as a primary second authentication factor. The lack of timescale, though, is a bit disappointing. Given the 2+ years time elapsed so far, I fear there will be a newsworthy account takeover of an account using security keys, via one of your currently enforced lower quality fallback methods, before GitHub becomes sufficiently focused on this.

**Reply by ar-an-ribe** [reply]
> Has Github dropped support for YubiKeys? The link on how to do this is now a dead link:
> 
> _[Here is](https://resources.github.com/whitepapers/GitHub-guide-to-commit-signing/) a detailed configuration guide for setting up your YubiKey with GitHub for commit verification and for SSH-based authentication._

The (PDF) configuration guide that was originally at that link is [downloadable from the Wayback Machine](https://web.archive.org/web/20210925073122/https://resources.github.com/downloads/GitHub-Whitepaper-Yubico-081021.pdf).

**Reply by 1larity** [reply]
A hardware key does not give microsoft access to your mobile phone as a permanent advertising identifier, this makes implementing hardware key support a lower priority than watering the plants in Sat Nad's office.

**Comment by Explorer09** [comment]
Dear GitHub,
Please allow me to set up Security Keys as an exclusive second factor for 2FA! I don't want any app installed for 2FA, or disclosing my phone number with SMS. Just allow me to set up Security Keys exclusively!!

Since you guys are going to force 2FA to many users (https://www.bleepingcomputer.com/news/software/github-warns-users-to-enable-2fa-before-upcoming-deadline/), it is kind of annoying to me that your site doesn't support a 2FA that best respect the user's privacy!

GitLab have done it right already. There is no reason for you to not do it.

**Comment by PyrateAkananto** [comment]
In case you own Yubikeys you can use the "Yubico Authenticator" app (in my case the Windows desktop app) to complete the whole TOTP process for the initial 2FA setup at GitHub and store everything related to it (except GitHub's recovery codes) on a Yubikey.

I did that myself and _then_ additionally added the same Yubikey as a "security key" to my GitHub account.

To sign in to GitHub I can now _either_ insert and press the YubiKey _or_ I can insert the Yubikey, launch the Yubico Authenticator app, read the TOTP from the app and enter it on GitHub.

**Reply by Explorer09** [reply]
I expect no apps be used when I set up a YubiKey as a 2FA. Not even the Yubico Authenticator app. Reason: They are black boxes with no way to tell if they secretly send out your TOTP secret to someone else. The reason I bought YubiKeys is for 2FA that's air-gapped. No spying; no proprietary code; no backdoor.

**Reply by clwgh** [reply]
@PyrateAkananto yeah that's a reasonable workaround for the current situation.

**Reply by PyrateAkananto** [reply]
@Explorer09 In my opinion using software from the manufacturer of a hardware key is still _a lot_ better than using a smartphone app or even unencrypted SMS. And if I already trust the hardware key of a certain manufacturer (which you also seem to do) it is no big step to trust the software of the very same manufacturer. Of course I would appreciate if GitHub let me just add hardware keys without forcing me to add something I absolutely do not need and do not want to use (TOTP, SMS).

**Reply by Explorer09** [reply]
@PyrateAkananto Actually I don't trust any software from the hardware key manufacturer. No manufacturer deserves trust. I trust the key hardware only because of its design -- (1.) no special driver or software needed for authentication; (2.) the keys stay offline until I need it for authentication (so the keys can't spy on me).

If I would need a TOTP app, I'll get an open source one (just for better trust), but it's still preferable if no apps are needed.

**Reply by netboy3** [reply]
> Reason: They are black boxes with no way to tell if they secretly send out your TOTP secret to someone else

@Explorer09 Yubico Authenticator does not have access to the secrets. It just sends the current time to the key and receives a code from the key. Also, the Yubico Authenticator is not a "black box" the code is [open source](https://github.com/Yubico/yubioath-flutter) and if you're concerned with it you are welcome to audit and/or build it yourself.

**Reply by Explorer09** [reply]
@netboy3 The average Android users cannot set up a build environment and build code themselves. The fact that Yubico Authenticator is open source doesn't help; users have to trust the official builds on Google Play anyway. The ideal 2FA is that users shouldn't assume trust on any app they installed. In other words, it should be a no-app solution.

**Reply by netboy3** [reply]
@Explorer09, your trust model doesn't make sense. The "average user" you relate to doesn't use open-source and personally built OS's either. They use Windows desktop and Android/iOS mobile devices that are all closed source. As mentioned previously, Yubikey does not expose TOTP secrets and computes the codes internally, on-key. This means that both the OS and the app have the same access level to generating codes and no access to secrets. What then would make a closed-source OS acceptable and an open-source application not acceptable? How would using an open-source app increase your risk level when the closed-source OS has the exact same access level to the resource?

**Reply by Explorer09** [reply]
@netboy3 There is no proof when an authenticator app sets up a secret to a key, the app won't keep a copy of the secret and covertly send it to some kind of government agency. When you set up a key for TOTP, the "secret" isn't generated by the key but the website/authentication servers. And the secret could leak to anywhere because the _protocol_ itself is weak. The need to install a mobile app is just a second argument against the TOTP. Just allow me to use U2F exclusively so I can get rid of all of the issues, please.

**Comment by candymappr** [comment]
Also used winauth to set up originally.

**Comment by lyronctk** [comment]
Two years later, still can't make a security key the sole authentication method of my github account.

Absolutely unacceptable for an organization that houses this much mission critical code.

**Reply by lyronctk** [reply]
Moving my engineering team over to Gitea.

**Comment by kevinforrestconnors** [comment]
There appears to be no way to remove TOTP over SMS from my 2FA configuration.  I want to only use Ubikey.  I also do not want to use recovery codes as if someone gains access to them, my account is completely compromised.

**Comment by CPU-Blanc** [comment]
Over 2 years on and we _still_ don't have support for only security keys for 2FA...
As others have said, I have security keys, but you're only as secure as your least secure login method, rendering the security keys next to useless. _Please_ let me disable TOTP from my account; offer it as an _option_, but don't force it on us if we want to use a more secure method...

**Comment by YoshiRulz** [comment]
Ah, I see this is a known issue.

I'll add that the live documentation is misleading.
[The screenshot here](https://docs.github.com/en/authentication/securing-your-account-with-two-factor-authentication-2fa/changing-your-two-factor-authentication-method#changing-the-totp-app) shows an account with 2 enabled 2FA methods and there's an option to remove one. I don't even see that dropdown, it's just an "Edit" button.
The [passkey setup](https://docs.github.com/en/authentication/securing-your-account-with-two-factor-authentication-2fa/configuring-two-factor-authentication#configuring-two-factor-authentication-using-a-passkey) contradictingly suggests "You must have already configured 2FA via a TOTP mobile app or via SMS," but also "If you use two-factor authentication (2FA), passkeys satisfy both password and 2FA requirements, so you can complete your sign in with a single step."
Buried on [the mandatory 2FA docs](https://docs.github.com/en/authentication/securing-your-account-with-two-factor-authentication-2fa/about-mandatory-two-factor-authentication#about-required-2fa-methods)—which is *not* where the reminder banner links to—you can find "Currently, we don't support passkeys or security keys as primary 2FA methods since they are easy to lose and do not support sync across a wide enough range of devices."

**Reply by MasterInQuestion** [reply]
"easy to lose and do not support sync"
    Sheer absurdity.
    Then what's the whole 2FA theater for..?

    Compared to that "easy to sync but also easy to lose/compromise" passwords:
    What "security" does it provide?

**Reply by DNin01** [reply]
>     "easy to lose and do not support sync"
> 
>     Sheer absurdity.
> 
>     Then what's the whole 2FA theater for..?
> 
> 
> 
>     Compared to that "easy to sync but also easy to lose/compromise" passwords:
> 
>     What "security" does it provide?
> 
> 

A lot of people lose their 2FA credentials, so GitHub has to take that into account, especially since they're making people who wouldn't otherwise adopt 2FA.

Passwords are a little different because it's common knowledge that you have to remember them and that you can usually reset them with an email.

Remember that any form of 2FA is better than not using 2FA, so use what's available while we wait for GitHub to hopefully make security keys a primary option.

**Reply by MasterInQuestion** [reply]
That appears to be a common misconception.
    Trace [ [https://github.com/orgs/community/discussions/63813#discussioncomment-10626201](https://github.com/orgs/community/discussions/63813#discussioncomment-10626201) ] for various details.

**Comment by imtiazvive** [comment]
Unfortunately, GitHub currently doesn’t allow you to fully disable text message or TOTP authentication while keeping only security keys for 2FA.

As per their documentation, authentication with a security key is considered secondary to TOTP or text message, which means you cannot entirely remove those options if you want to use a security key as your primary method. 

While you can keep recovery codes as a backup, you'll still need to rely on TOTP or text messages for 2FA at this time.

**Comment by okarmazin** [comment]
What is the point of allowing security keys if you then still require the less secure TOTP? 

The reasoning of GitHub staff is completely backwards. THE ENTIRE POINT of setting up a physical security key is that it CANNOT be shared to a random scammer, unlike a TOTP code.

Am I crazy? Why would you use the "cannot be shared" as a negative reason to keep requiring the shareable 2FA method?

**Comment by s1ugh34d** [comment]
So I can't use my secure hardware keys, but I have to buy a cell phone plan to login to github? Let me know when you develop real security

**Reply by Explorer09** [reply]
@s1ugh34d I hate when GitHub does nothing on improving 2FA security, but here are workarounds I found: *You don't need a cell phone for 2FA*.

* Authenticator apps for desktop: There are many authenticator apps for desktop (PCs) that support the same TOTP as Microsoft Authenticator and Google Authenticator. Since I don't like Google and Microsoft spying on users. A recommended alternative is [Yubico Authenticator](https://www.yubico.com/products/yubico-authenticator/).
* If you want a very minimal implementation of TOTP, which supports no credential management at all, [OATH Toolkit](https://oath-toolkit.codeberg.page/) would be the tool for you. It's FOSS, command line only, and available for Linux, macOS and FreeBSD. (Can also work with Windows under Cygwin, but you need to build it yourself.)
* If you have a YubiKey (full featured, not Security Key by Yubico that supports FIDO only), you can store the TOTP secret into the key. Use [YubiKey Manager CLI](https://developers.yubico.com/yubikey-manager/) to store secret to the key and generate OTP from it. The commands are [`ykman oath accounts add` and `ykman oath accounts code`](https://docs.yubico.com/software/yubikey/tools/ykman/OATH_Commands.html#ykman-oath-accounts-options-command-args).

**Comment by Explorer09** [comment]
I hate when GitHub does nothing on improving 2FA security, but here are workarounds I found: *You don't need a cell phone for 2FA*.

* Authenticator apps for desktop: There are many authenticator apps for desktop (PCs) that support the same TOTP as Microsoft Authenticator and Google Authenticator. Since I don't like Google and Microsoft spying on users. A recommended alternative is [Yubico Authenticator](https://www.yubico.com/products/yubico-authenticator/).
* If you want a very minimal implementation of TOTP, which supports no credential management at all, [OATH Toolkit](https://oath-toolkit.codeberg.page/) would be the tool for you. It's FOSS, command line only, and available for Linux, macOS and FreeBSD. (Can also work with Windows under Cygwin, but you need to build it yourself.)
* If you have a YubiKey (full featured, not Security Key by Yubico that supports FIDO only), you can store the TOTP secret into the key. Use [YubiKey Manager CLI](https://developers.yubico.com/yubikey-manager/) to store secret to the key and generate OTP from it. The commands are [`ykman oath accounts add` and `ykman oath accounts code`](https://docs.yubico.com/software/yubikey/tools/ykman/OATH_Commands.html#ykman-oath-accounts-options-command-args).

**Comment by s1ugh34d** [comment]
I instead of making my own workarounds for a multi million dollar companies
failings, I deleted my github account and switched to tangled.org

s1ugh34d

The best lack all conviction, while the worst are full of passionate
intensity.
      -W. B. Yeats

It’s a dangerous thing to mistake speaking without thought for speaking the
truth.
      -Benoit Blanc

Failure is not an option — it comes bundled with Windows.

**Comment by gerethhart** [comment]
It's 2026 and this still is not available. This seems more like unwillingness to do it, than any technical limitation. Requiring TOTP is forcing a relatively insecure method compared to properly implemented hardware security keys + confirmation code.

I know the capability is there because you enterprise offerings in the gov space are capable of this via PIV/CAC cards.

**Reply by PhilienTaylor** [reply]
I was literally looking up apps that do this and I'm shocked it's not available for GitHub. People sign commits with gpg keys all day and we can't sign in using FIDO2? make it make sense.
